Add Concurnas support

jasontatton opened the pull request:
Concurnas is an open source JVM programming language designed for building reliable, scalable, high performance concurrent, distributed and parallel systems. For more information please see: https://concurnas.com/

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,38 @@
+		greedy: true
+	},
+	'function': {
+		pattern: /((?:^|\s)def[ \t]+)[a-zA-Z_]\w*(?=\s*\()/g,
```
The `g` flag isn't necessary.
```suggestion
		pattern: /((?:^|\s)def[ \t]+)[a-zA-Z_]\w*(?=\s*\()/,
```

jasontatton replied:
done, thanks!

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,38 @@
+		pattern: /((?:^|\s)def[ \t]+)[a-zA-Z_]\w*(?=\s*\()/g,
+		lookbehind: true
+	},
+	'keyword': /\b(?:class|enum|annotation|def|init|gpudef|gpukernel|typedef|int|size_t|long|short|float|double|boolean|bool|byte|char|lambda|void|import|from|using|assert|inject|provider|provide|single|nodefault|unchecked|global|local|constant|private|public|package|protected|true|false|null|this|super|del|await|trans|transient|shared|new|abstract|override|extends|trait|val|var|in|out|default|sizeo ...
```
Please sort this keyword list.

jasontatton replied:
done, thanks!

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,38 @@
+	'number': /\b0b[01][01_]*L?\b|\b0x[\da-f_]*\.?[\da-f_p+-]+\b|(?:\b\d[\d_]*\.?[\d_]*|\B\.\d[\d_]*)(?:e[+-]?\d[\d_]*)?[dfl]?/i,
+	'punctuation': /[{}[\];(),.:]/,
+	'operator': {
+		pattern: /<==|>==|=>|->|<-|<>|\^|&==|&<>|!|\?|\?:|\.\?|\+\+|--|[-+*/=<>]=?|\b(?:as|is|isnot|bor|band|bxor|comp|mod|and|inv|or|shl|shr|ushr|xor)\b=?/
```
1. You can inline the regex: `'operator': /<==|.../`
2. Aren't the following also operators? `~`, `&`, `@`, `:`, `:=`

jasontatton replied:
I have inlined as requested. ~, &,  :, := are not operators but @ can be (used in annotations and is the copy operator)

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,38 @@
+
+};
+
+Prism.languages.insertBefore('concurnas', 'keyword', {
```
Instead of using `insertBefore` you can just insert it before `keyword` yourself.

`insertBefore` is usually only necessary with cloned languages from `Prism.languages.extend`.

jasontatton replied:
done, thanks!

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,38 @@
+	},
+	'keyword': /\b(?:class|enum|annotation|def|init|gpudef|gpukernel|typedef|int|size_t|long|short|float|double|boolean|bool|byte|char|lambda|void|import|from|using|assert|inject|provider|provide|single|nodefault|unchecked|global|local|constant|private|public|package|protected|true|false|null|this|super|del|await|trans|transient|shared|new|abstract|override|extends|trait|val|var|in|out|default|sizeo ...
+	'boolean': /\b(?:true|false)\b/,
+	'number': /\b0b[01][01_]*L?\b|\b0x[\da-f_]*\.?[\da-f_p+-]+\b|(?:\b\d[\d_]*\.?[\d_]*|\B\.\d[\d_]*)(?:e[+-]?\d[\d_]*)?[dfl]?/i,
```
I don't think this works for [`short` literals](https://concurnas.com/docs/types.html#primitive-types) (e.g. 16s), does it?

jasontatton replied:
fixed, thanks!

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,38 @@
+		}
+	],
+	'string': {
+		pattern: /(?:[rub]|rb|br)?(\|\||r"|r'|"|')(?:\\.|(?!\1)[^\\\r\n])*\1/i,
```
So `rbr"r"` a valid string?

In the [string section](https://concurnas.com/docs/types.html#strings), I could only find `r"|r'|"|'` and `s"|s'` which isn't supported. Is `||foo||` really valid?
Also, is the `r`/`s` prefix really case-insensitive?
```suggestion
		pattern: /[rs]?("|')(?:\\.|(?!\1)[^\\\r\n])*\1/,
```

jasontatton replied:
Thanks for this. ||foo|| on it's own is not valid, but the likes of: mylang ||foo|| are. I have added a langext rule to support this. Also I have added code to support 'code in strings'

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,38 @@
+	'operator': {
+		pattern: /<==|>==|=>|->|<-|<>|\^|&==|&<>|!|\?|\?:|\.\?|\+\+|--|[-+*/=<>]=?|\b(?:as|is|isnot|bor|band|bxor|comp|mod|and|inv|or|shl|shr|ushr|xor)\b=?/
+	}
+
```
Minor nit. Please remove the empty line.

jasontatton replied:
done thanks

jasontatton commented:
Hello, thank you for taking the time to review this PR and especially for going into the documentation for Concurnas and pointing out a lot of things which I forgot/implemented incorrectly (btw I hope you like Concurnas as a language and I'd love to receive your advice on how it can be improved).

I have now fixed the things raised and a few more.

Bugfixes:
+ added support for code in strings.
+ remove g from function pattern
+ sorted the keyword list
+ inlined operator
+ removed unecisary use of insertBefore 
+ added short literal support
+ lots of fixes to Strings
+ Improved rule for langExtensions greatly - must be prefixed with the extension name
+ fixed operator list
+ removed whitespace

Also:
+ Improved tests and example.html

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,50 @@
+			'interpolation': {
+				pattern: /((?:^|[^\\])(?:\\{2})*){(?:[^{}]|{(?:[^{}]|{[^}]*})*})+}/,
+				lookbehind: true,
+				inside: {
```
You can just do `inside: Prism.languages.concurnas`.

jasontatton replied:
done, thanks!

RunDevelopment commented on components/prism-concurnas.js:
```diff
@@ -0,0 +1,50 @@
+		}
+	],
+	'langext': {
+		pattern: /(?:\w+)\s*(\|\|)[\s\S]+?\1/i,
```
```suggestion
		pattern: /\w+\s*\|\|[\s\S]+?\|\|/,
```

jasontatton replied:
done, thanks!

RunDevelopment commented:
Thank you for contributing @jasontatton!
